### PR TITLE
Scala 2.13

### PR DESCRIPTION
**Pull request by jackkoenig**
This is currently a draft PR, there are several things I want to do before merging:

TODO
- [x] ~Test with rocket-chip / other large projects~ Easier to do once this is part of a SNAPSHOT
- [x] Examine the `VecLike` change from `collection.IndexedSeq` to `IndexedSeq` (presumably `collection.immutable.IndexedSeq`)
- [x] See if we can change the type of `Record.elements` from `ListMap` to `SeqMap`, at least in Scala 2.13
- [x] ~Factor out `InternalDontCare` -> `DontCare` change and backport to 3.4 (see https://github.com/chipsalliance/chisel3/issues/1727)~ This would be a binary incompatible change so can't backport

### Contributor Checklist

- [ ] Did you add Scaladoc to every public function/method?
- [ ] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

  - performance improvement 
  - code refactoring 
  - new feature/API 

#### API Impact

Primarily an expansion adding cross-build for Scala 2.13.

Changes:
* `DontCare` is now a top-level object instead of a val pointing to an internal thing (This is source compatible but not binary compatible, Fixes https://github.com/chipsalliance/chisel3/issues/1727)
* `Record.elements` and `Bundle.elements` are now of type `SeqMap` (new in Scala 2.13). We include aliasing from `SeqMap` and `VectorMap` to `ListMap` in for 2.12. This aliasing does make the ScalaDoc slightly awkward unfortunately.
* `Record.elements` is now a `def` instead of a `val`. This is incidental but it helps reduce the memory use of some `Records` that may keep their elements in a different val and then just map to it for `elements`.
* `VecLike` now extends `collections.immutable.IndexedSeq` instead of `collections.IndexedSeq`. This mirrors the changes in Scala 2.13 of `Seq` now extending `collections.immutable.Seq` (instead of `collections.Seq`). For simplicity, this change affects the 2.12 build as well. It *shouldn't* affect any users, but it's something to keep in mind.

#### Backend Code Generation Impact

No impact

#### Desired Merge Strategy

   - Squash

#### Release Notes

* Add support for Scala 2.13. Please see the official [Scala 2.13.0 release notes](https://github.com/scala/scala/releases/tag/v2.13.0) for more information about the changes from Scala 2.12 to 2.13. For most users, this involves updating `scalaVersion` in your `build.sbt` file.
* `DontCare` is now a top-level object instead of an alias
* `Record.elements` and `Bundle.elements` are now of type `SeqMap` (new in Scala 2.13). They still are of type `ListMap` in Scala 2.12
* `Record.elements` is now a `def` instead of a `val`
* `VecLike` now extends `collections.immutable.IndexedSeq` instead of `collections.IndexedSeq`

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (3.2.x, 3.3.x, 3.4.0, 3.5.0) ?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you mark as `Please Merge`?

### CODE REVIEW

File: core/src/main/scala/chisel3/Data.scala [hunk 1]
```diff
@@ -743,6 +744,10 @@ object WireDefault {
     applyImpl(t, init)
   }
 
+  def foo[T <: Data](t: T, init: DontCare.type)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): T = {
+    applyImpl(t, init)
+  }
+
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/Data.scala:750]: Whoops I forgot I left this in, TODO delete

File: core/src/main/scala/chisel3/package.scala [hunk 2]
```diff
@@ -207,9 +207,6 @@ package object chisel3 {
     a.allElements
   }
   def getModulePorts(m: Module): Seq[Port] = m.getPorts
-  // Invalidate API - a DontCare element for explicit assignment to outputs,
-  //  indicating the signal is intentionally not driven.
-  val DontCare = chisel3.internal.InternalDontCare
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/package.scala:212]: Something about this being a val caused type inference issues with the versions of things like [`WireDefault.apply` ](https://www.chisel-lang.org/api/3.4.1/chisel3/WireDefault$.html#apply[T%3C:chisel3.Data](t:T,init:chisel3.DontCare.type)(implicitsourceInfo:chisel3.internal.sourceinfo.SourceInfo,implicitcompileOptions:chisel3.CompileOptions):T)that accept `DontCare.type`. I don't know exactly what the change is, but the solution, to just get rid of the aliasing is cleaner anyway. It also solves https://github.com/chipsalliance/chisel3/issues

File: core/src/main/scala-2.12/scala/collection/immutable/package.scala [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+import scala.collection.immutable.ListMap
+
+package object immutable {
+  val SeqMap = ListMap
+  type SeqMap[K, +V] = ListMap[K, V]
+
+  val VectorMap = ListMap
+  type VectorMap[K, +V] = ListMap[K, V]
+}
```
Comment (jackkoenig) [inline core/src/main/scala-2.12/scala/collection/immutable/package.scala:13]: Here's a neat trick, we can use `SeqMap` and `VectorMap` within Chisel, but just alias them to regular old `ListMap` 🙂

Comment (seldridge) [reply core/src/main/scala-2.12/scala/collection/immutable/package.scala:13]:
```scala
object Quicksort {
  def apply(a: Seq[Int]) = BubbleSort(a)
}
```

**Comment by ekiwi** [comment]
Seems like the CI works :tada:!

Did you want to get this into the pre-release branch soon? Then I could start working on `2.13` support for `chiseltest`.

**Comment by jackkoenig** [comment]
> Did you want to get this into the pre-release branch soon? Then I could start working on 2.13 support for chiseltest.

Yeah, will need to try this with rocket-chip and internal projects but I think it's better to get this out and we can iterate on it before 3.5 release if needed.

**Review by ekiwi — APPROVED** [review]
LGTM

Please add the `PleaseMerge` tag yourself if you feel that it is ready.

**Pull request merged**